File: sdks/suprawall-java/src/main/java/io/suprawall/SupraWall.java
package io.suprawall;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

public class SupraWall {
    private static String apiKey = System.getenv("SUPRAWALL_API_KEY");
    private static String apiUrl = System.getenv().getOrDefault("SUPRAWALL_API_URL", "https://api.suprawall.io/v1/evaluate");

    public static void setApiKey(String key) { apiKey = key; }
    public static void setApiUrl(String url) { apiUrl = url; }

    public static class PolicyViolationException extends Exception {
        public PolicyViolationException(String msg) { super(msg); }
    }

    public static String evaluate(String agentId, String toolName, String jsonArguments) throws PolicyViolationException, IOException, InterruptedException {
        if (apiKey == null) throw new IllegalArgumentException("SUPRAWALL_API_KEY is required");

        if (apiKey.startsWith("ag_test_")) {
            return "{\"decision\":\"ALLOW\",\"reason\":\"Test mode bypass\"}";
        }

        String jsonPayload = String.format("{\"agentId\":\"%s\",\"toolName\":\"%s\",\"arguments\":%s}", agentId, toolName, jsonArguments == null ? "{}" : jsonArguments);

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl))
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonPayload))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 200) {
           throw new IOException("SupraWall Network Error: Failing closed");
        }

        String body = response.body();
        if (body.contains("\"decision\":\"DENY\"")) {
            throw new PolicyViolationException("SupraWall Policy Violation: Tool '" + toolName + "' explicitly denied.");
        } else if (body.contains("\"decision\":\"REQUIRE_APPROVAL\"")) {
            throw new PolicyViolationException("SupraWall Policy Violation: Tool '" + toolName + "' requires human approval.");
        }

        return body;
    }
}

File: packages/sdk-ruby/lib/suprawall.rb
require 'net/http'
require 'json'
require 'uri'

module SupraWall
  class Error < StandardError; end
  class PolicyViolationError < Error; end

  DEFAULT_API_URL = 'https://www.supra-wall.com/api/v1/evaluate'

  class << self
    attr_accessor :api_key
    attr_accessor :api_url

    def setup
      yield self
    end

    def evaluate!(agent_id:, tool_name:, arguments: {})
      api_key = self.api_key || ENV['SUPRAWALL_API_KEY']
      api_url = self.api_url || ENV['SUPRAWALL_API_URL'] || DEFAULT_API_URL

      raise Error, "SUPRAWALL_API_KEY is required" unless api_key

      # Bypass completely if it's a test key
      return { "decision" => "ALLOW", "reason" => "Test mode" } if api_key.start_with?("sw_test_")

      uri = URI(api_url)
      req = Net::HTTP::Post.new(uri)
      req['Authorization'] = "Bearer #{api_key}"
      req['Content-Type'] = 'application/json'
      req.body = {
        agentId: agent_id,
        toolName: tool_name,
        arguments: arguments
      }.to_json

      res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https') do |http|
        http.request(req)
      end

      unless res.is_a?(Net::HTTPSuccess)
        raise Error, "SupraWall Network Error: Failing closed."
      end

      data = JSON.parse(res.body)
      
      if data['decision'] == 'DENY'
        raise PolicyViolationError, "SupraWall: Tool '#{tool_name}' explicitly denied."
      elsif data['decision'] == 'REQUIRE_APPROVAL'
        raise PolicyViolationError, "SupraWall: Tool '#{tool_name}' requires human approval."
      end

      data
    end
  end
end
